tests/ga/test_remoteaccess_handler.py

```diff
@@ -0,0 +1,358 @@
+    def test_handle_remote_access_user_removed(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        data_str = load_data('wire/remote_access_10_accounts.xml')
+        remote_access = RemoteAccess(data_str)
+        count = 0
+        for user in remote_access.user_list.users:
+            count += 1
+            user.name = "tstuser{0}".format(count)
+            expiration_date = datetime.utcnow() + timedelta(days=count)
+            user.expiration = expiration_date.strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        rah.remote_access = remote_access
+        rah.handle_remote_access()
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertEqual(10, os_util.all_users.keys().__len__())
+        del rah.remote_access.user_list.users[:]
+        self.assertEqual(10, os_util.all_users.keys().__len__())
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_handle_remote_access_bad_data_and_good_data(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        data_str = load_data('wire/remote_access_10_accounts.xml')
+        remote_access = RemoteAccess(data_str)
+        count = 0
+        for user in remote_access.user_list.users:
+            count += 1
+            user.name = "tstuser{0}".format(count)
+            if count is 2:
+                user.name = ""
+            expiration_date = datetime.utcnow() + timedelta(days=count)
+            user.expiration = expiration_date.strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        rah.remote_access = remote_access
+        rah.handle_remote_access()
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertEqual(9, os_util.all_users.keys().__len__())
+
+    @patch('azurelinuxagent.common.utils.cryptutil.CryptUtil.decrypt_secret',
+    return_value="]aPPEv}uNg1FPnl?")
+    def test_handle_remote_access_deleted_user_readded(self, _):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        data_str = load_data('wire/remote_access_single_account.xml')
+        remote_access = RemoteAccess(data_str)
+        tstuser = remote_access.user_list.users[0].name
+        expiration_date = datetime.utcnow() + timedelta(days=1)
+        expiration = (expiration_date).strftime("%a, %d %b %Y %H:%M:%S ") + "UTC"
+        remote_access.user_list.users[0].expiration = expiration
+        rah.remote_access = remote_access
+        rah.handle_remote_access()
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users".format(tstuser))
+        os_util.all_users.clear()
+        self.assertTrue(tstuser not in os_util.all_users)
+        rah.handle_remote_access()
+        self.assertTrue(tstuser in os_util.all_users, "{0} missing from users".format(tstuser))
```

Comment:
You should be testing the behavior of the architected interface (os_util.get_users()).

Reply:
Changed

---
In reply to: [194544049](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r194544049) [](ancestors = 194544049)